src/Psalm/Internal/Analyzer/ClassAnalyzer.php

```diff
@@ -2627,5 +2635,119 @@ private function checkEnum(PhpParser\Node\Stmt\Enum_ $enum): void
                 }
             }
         }
+
+        $case_values = [];
+        $case_names = [];
+
+        foreach ($enum->stmts as $stmt) {
+            if ($stmt instanceof PhpParser\Node\Stmt\EnumCase) {
+                if (isset($case_names[$stmt->name->name])) {
+                    if (IssueBuffer::accepts(
+                        new DuplicateEnumCase(
+                            'Enum case names should be unique',
+                            new CodeLocation($this, $stmt),
+                            $enum->name->name
+                        )
+                    )) {
+                        continue;
+                    }
+                }
+
+                $case_names[$stmt->name->name] = true;
+
+                if (null !== $stmt->expr && null === $enum->scalarType) {
+                    if (IssueBuffer::accepts(
+                        new InvalidEnumCaseValue(
+                            'Case of a non-backed enum should not have a value',
+                            new CodeLocation($this, $stmt),
+                            $enum->name->name
+                        )
+                    )) {
+                        continue;
+                    }
+                } elseif (null === $stmt->expr && null !== $enum->scalarType) {
+                    if (IssueBuffer::accepts(
+                        new InvalidEnumCaseValue(
+                            'Case of a backed enum should have a value',
+                            new CodeLocation($this, $stmt),
+                            $enum->name->name
+                        )
+                    )) {
+                        continue;
+                    }
+                } elseif (null !== $enum->scalarType && null !== $stmt->expr) {
+                    if (ExpressionAnalyzer::analyze(
+                        new StatementsAnalyzer(
+                            $this,
+                            $type_provider = new \Psalm\Internal\Provider\NodeDataProvider()
+                        ),
+                        $stmt->expr,
+                        $enum_context,
+                        false,
+                        $global_context,
+                        true
+                    ) === false) {
+                        // xxx: should we rather return here?
```

Comment:
I think we should. false returns happen when things go pretty wrong,  like UndefinedClass wrong. When that happens, it's no use trying to understand what's going on for the rest of the scope